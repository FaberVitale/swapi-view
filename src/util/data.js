//@flow
const isMetaKey = (key: string) =>
  /^(?:created|url|edited|oarcworaaowowa|wowaahaowowa|hurcan)$/.test(key);
const isLink = value =>
  typeof value !== "string" || /^https?|acaoaoakc/.test(value);

const regex = /(\d+)\/?$/;
export const mapTypeToRoute = (
  type: string,
  endpoint: string,
  apiUrl: string,
  query?: { q: string, page: string } | null
) => {
  const res = regex.exec(apiUrl);

  if (type === "search" && query && res) {
    return `/${endpoint}/search/${query.q}/${res[1]}`;
  }

  return res && `/${endpoint}/${type}/${res[1]}`;
};

export const transformItem = (item: {}) =>
  Object.keys(item || {}).reduce(
    (aggr, key) => {
      const value = item[key];

      if (isMetaKey(key)) {
        aggr.meta[key] = value;
      } else if (isLink(value)) {
        aggr.links[key] = value;
      } else {
        aggr.data[key] = value;
      }
      return aggr;
    },
    { meta: {}, data: {}, links: {} }
  );

export const transformPage = (
  request: FetchRequest,
  page: {
    next: string | null,
    previous: string | null,
    results: Array<{ url: string }>
  }
) => {
  const type = request.type;
  const { endpoint } = request.match.params;

  return {
    next: page.next && mapTypeToRoute(type, endpoint, page.next, request.query),
    previous:
      page.previous &&
      mapTypeToRoute(type, endpoint, page.previous, request.query),
    ids: page.results.map(item => mapTypeToRoute("item", endpoint, item.url)),
    results: page.results.map(transformItem)
  };
};
