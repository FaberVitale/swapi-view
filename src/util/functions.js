//@flow
import _get from "lodash.get";

export const getCurried = (keys: Array<string>, or: any) => (val: Object) =>
  _get(val, keys, or);

export const identity = (id: mixed) => id;

export const notEqual = (a: mixed) => (b: mixed) => a !== b;

export const equals = (a: mixed) => (b: mixed) => a === b;

export const isTruthy = (val: mixed) => !!val;

export const equalsTo = (...args: Array<mixed>) => (val: mixed) =>
  args.some(equals(val));

export const getDisplayName = (hocName: string, Comp: any) =>
  `with${hocName}(${Comp.displayName || Comp.name || "Component"})`;

export const warn: (...msgs: Array<any>) => void =
  process.env.NODE_ENV !== "production" &&
  window.console &&
  typeof window.console.warn === "function"
    ? window.console.warn
    : () => {};
