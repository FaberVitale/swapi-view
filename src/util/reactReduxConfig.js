//@flow
import { fetchRouteResource } from "../action/creators";
import { getData } from "../selectors/data";
import { getFetchState } from "../selectors/fetchState";
import type { Match, Location } from "react-redux";
import type { State } from "../reducers";
import type { Dispatch } from "redux";
import { connect } from "react-redux";
import { isDrawerOpen } from "../selectors/drawer";
import { drawerRequestChange } from "../action/creators";

type DispatchProps = {
  load: () => void
};

type StateProps = {};

type RouterProps = {
  match: Match,
  location: Location,
  history: History
};

/* these are the props passed to Page and Item */
const mergePropsView: (
  stateProps: StateProps,
  dispatchProps: DispatchProps,
  ownProps: RouterProps
) => * = (stateProps, dispatchProps, ownProps) => {
  let res = {};

  res.match = ownProps.match;
  res.search = ownProps.location.search;
  return Object.assign(res, stateProps, dispatchProps);
};

/* the function load is used to request data from api on location.pathname changes*/
export const mapDispatchToPropsView = (
  dispatch: Dispatch<Action>,
  ownProps: RouterProps
) => ({
  load: () => {
    dispatch(
      fetchRouteResource(ownProps.match, ownProps.location.search, false)
    );
  }
});
// re-render for changes in state.data or if the url changes
const connectOptionsView = {
  pure: true,
  areStatesEqual: (next: State, prev: State) =>
    getData(next) === getData(prev) &&
    getFetchState(next) === getFetchState(prev),
  areOwnPropsEqual: (next: RouterProps, prev: RouterProps) =>
    next.location.pathname === prev.location.pathname &&
    next.location.search === prev.location.search
};

export const connectView = (
  mapStateToProps: ?(state: State, ownProps: RouterProps) => Object = null
) => {
  return connect(
    mapStateToProps,
    mapDispatchToPropsView,
    mergePropsView,
    connectOptionsView
  );
};

export const mapStateToPropsDrawer = (state: State) => ({
  isOpen: isDrawerOpen(state)
});

// we need to throttle because material buttons trigger twice  event listeners
// applied to a "click" event, if Enter is input
// see: https://github.com/mui-org/material-ui/issues/9344
export const mapDispatchToPropsDrawer = (dispatch: Dispatch<Action>) => ({
  toggleDrawer: (value: boolean) => dispatch(drawerRequestChange(value))
});

export const connectDrawer = (addDispatchTopProps?: boolean = false) => {
  return connect(
    mapStateToPropsDrawer,
    addDispatchTopProps ? mapDispatchToPropsDrawer : null,
    undefined,
    {
      pure: true,
      areStatesEqual: (next, prev) => isDrawerOpen(next) === isDrawerOpen(prev)
    }
  );
};
