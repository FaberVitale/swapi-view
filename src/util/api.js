//@flow
import { warn } from "./functions";

/* Thrown when a GET request returns with a status !== 200.
 * based on https://stackoverflow.com/a/5251506
 */
export function ApiError(url: string, status: number) {
  this.url = url;
  this.status = status;
  this.name = "ApiError";
  this.message = `url: ${url}, status: ${status}`;
  this.stack = new Error().stack;
}
// $FlowFixMe
ApiError.prototype = new Error();

/* higher order function that given an optional timeout
  returns a fetch-like function that
  on fulfill will returns the response,
  on rejection returns the following errors:
  
  1 - TypeError: Network error or Permission error
  2 - ApiError: If the response status isnt 200
  3 - SyntaxError: Invalid JSON
  4 - an Error with message "timeout" if the request is hanging
  5 - RangeError - If the response status isnt in range [200; 599]

  spec: https://fetch.spec.whatwg.org/
*/
export const fetchData = (timeout?: number = 10000) => {
  let wait: number;

  if (!Number.isSafeInteger(timeout) || timeout < 500) {
    warn(
      `timeout is not a valid number:
       expected an Integer in range [500, 2^31 -1], got: ${timeout}
       timeout set to 10s`
    );
    wait = 10000;
  }
  wait = timeout;

  return async (url: string | URL, init?: RequestOptions) => {
    let res;

    res = await new Promise((resolve, reject) => {
      let tRef = setTimeout(() => {
        reject(new Error("timeout"));
      }, wait);

      const req = init ? fetch(url, init) : fetch(url);

      req.then(response => {
        clearTimeout(tRef);
        resolve(response);
      }, reject);
    });

    if (res.status === 200) {
      return res;
    }

    throw new ApiError(
      typeof url !== "string" ? url.toString() : url,
      res.status
    );
  };
};
