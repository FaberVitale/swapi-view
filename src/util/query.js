//@flow
import qs from "qs";
import { queryRegex, qsOpt } from "../routes";
import { warn } from "../util/functions";

/* returns the string query if present, null otherwise
 * Note: it doesnt support array params
 */
export const qsStringify = (arg?: {} = {}) => qs.stringify(arg, qsOpt);

export const getQueryValue = (parsed: {}, key: string, regex?: RegExp) => {
  let value: string;
  let temp: string | Array<string> = parsed[key] || "";

  if (!temp || typeof temp !== "string") {
    return null;
  }

  value = temp;

  if (regex) {
    return regex.test(value) ? value : null;
  }

  return value;
};

//parses and validates the query using queryRegex
export const getQuery = (search: string) => {
  let parsed, query;

  if (typeof search !== "string" || !search) {
    return null;
  }

  query = search[0] === "?" ? search.slice(1) : search;

  try {
    parsed = qs.parse(query);
  } catch (e) {
    warn(e);
  }

  if (parsed) {
    const q = getQueryValue(parsed, "q", queryRegex.q) || "";
    const page = getQueryValue(parsed, "page", queryRegex.page) || "1";

    if (q) {
      return {
        q,
        page
      };
    }
  }

  return null;
};

//returns the pathname + search part of the url  of the result query
export const mapQueryToName = (
  endpoint: string,
  query: { q: string, page: string } | null
) => {
  if (query) {
    const stringified = qsStringify(query);

    return `/${endpoint}/search?${stringified}`;
  } else {
    return `/${endpoint}/search`;
  }
};
