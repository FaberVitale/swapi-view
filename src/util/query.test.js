import { getQueryValue, getQuery, mapQueryToName } from "./query";
import { equals } from "../util/functions";

describe("util/query", () => {
  const parsed = { q: "20", arr: ["40", "50"] };

  describe("query/getQUeryValue", () => {
    it("returns null if not present", () => {
      expect(getQueryValue(parsed, "notPresent")).toBe(null);
      expect(getQueryValue(parsed, "notPresent", /\w+/)).toBe(null);
    });

    it("returns the value, if present and regex is not passed", () => {
      expect(getQueryValue(parsed, "q")).toBe("20");
    });

    it("returns the value, if present and if matches the regex", () => {
      expect(getQueryValue(parsed, "q", /\d{1,2}/)).toBe("20");
    });

    it("returns null, if present but doesnt not match the regex", () => {
      expect(getQueryValue(parsed, "q", /[A-Z]+/)).toBe(null);
    });

    test("if key is points to an array, returns null", () => {
      expect(getQueryValue(parsed, "arr")).toBe(null);
    });
  });

  describe("query/getQuery", () => {
    it("returns null if search is not string or is an empty string", () => {
      expect(
        [null, undefined, 0, /ab/, ""].map(r => getQuery(r)).every(equals(null))
      ).toBe(true);
    });

    it("processes correctly string with question mark", () => {
      expect(getQuery("?q=asd&page=2")).toEqual({
        q: "asd",
        page: "2"
      });
    });

    it("returns null if it doesnt contain q", () => {
      expect(getQuery("page=20")).toBe(null);
    });

    it("sets page to 1 if it is not provided", () => {
      expect(getQuery("q=asdf")).toEqual({
        q: "asdf",
        page: "1"
      });
    });

    it("returns the correct query object, if a valid query string is provided", () => {
      expect(getQuery("page=3&q=etgc")).toEqual({
        q: "etgc",
        page: "3"
      });
    });
  });

  describe("query/mapQueryToName", () => {
    it("works correctly", () => {
      expect(mapQueryToName("people", { q: "a", page: "2" })).toMatchSnapshot();

      expect(mapQueryToName("people", null)).toMatchSnapshot();
    });
  });
});
