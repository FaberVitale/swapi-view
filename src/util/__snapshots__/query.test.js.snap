// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`util/query query/mapQueryToName works correctly 1`] = `"/people/search?q=a&page=2"`;

exports[`util/query query/mapQueryToName works correctly 2`] = `"/people/search"`;
